modules/contracts/apps/subscription/src/handlers/execute.rs

```diff
@@ -0,0 +1,351 @@
+use crate::contract::{SubscriptionApp, SubscriptionResult};
+use crate::msg::{SubscriptionExecuteMsg, UnsubscribedHookMsg};
+use crate::state::{
+    EmissionType, Subscriber, SubscriptionConfig, SubscriptionState, EXPIRED_SUBSCRIBERS,
+    INCOME_TWA, SUBSCRIBERS, SUBSCRIPTION_CONFIG, SUBSCRIPTION_STATE,
+};
+use crate::{SubscriptionError, DURATION_IN_WEEKS, WEEK_IN_SECONDS};
+use abstract_sdk::{AbstractResponse, Execution, ExecutorMsg, TransferInterface};
+use cosmwasm_std::{Addr, Decimal, Deps, DepsMut, Env, MessageInfo, Response, StdResult, Uint128};
+use cw_asset::{Asset, AssetInfoUnchecked};
+
+pub fn execute_handler(
+    mut deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    app: SubscriptionApp,
+    msg: SubscriptionExecuteMsg,
+) -> SubscriptionResult {
+    match msg {
+        SubscriptionExecuteMsg::Pay { subscriber_addr } => {
+            let maybe_received_coin = info.funds.last();
+            let subscriber_addr = subscriber_addr
+                .map(|human| deps.api.addr_validate(&human))
+                .transpose()?
+                .unwrap_or(info.sender.clone());
+            if let Some(coin) = maybe_received_coin.cloned() {
+                try_pay(app, deps, env, info, Asset::from(coin), subscriber_addr)
+            } else {
+                Err(SubscriptionError::NotUsingCW20Hook {})
+            }
+        }
+        SubscriptionExecuteMsg::Unsubscribe { unsubscribe_addrs } => {
+            unsubscribe(deps, env, app, unsubscribe_addrs)
+        }
+        SubscriptionExecuteMsg::ClaimEmissions { addr } => {
+            claim_subscriber_emissions(&app, &mut deps, &env, addr)
+        }
+        SubscriptionExecuteMsg::UpdateSubscriptionConfig {
+            payment_asset,
+            subscription_cost_per_week: subscription_cost,
+            subscription_per_week_emissions,
+            unsubscription_hook_addr,
+        } => update_subscription_config(
+            deps,
+            env,
+            info,
+            app,
+            payment_asset,
+            subscription_cost,
+            subscription_per_week_emissions,
+            unsubscription_hook_addr,
+        ),
+        SubscriptionExecuteMsg::RefreshTWA {} => {
+            INCOME_TWA.try_update_value(&env, deps.storage)?;
+            Ok(Response::new())
+        }
+    }
+}
+
+/// Called when either paying with a native token or through the receive_cw20 endpoint when paying
+/// with a CW20.
+pub fn try_pay(
```

Comment:
Might just be me but I think this function could benefit from refactoring some of the inner logic into their own functions to improve readability.

Reply:
Did some basic refactor to this method! Do you think it's enough for now?